### PR TITLE
fix(push): send push control events to Divine relay with relay OK handling

### LINKED ISSUE #4411: fix(push): send push control events to Divine relay with relay OK handling
## Summary
Mobile push control events need to reliably reach the Divine relay path consumed by `divine-push-service` so real FCM push registration, deregistration, and preference updates work in production.

## Context
Recent investigation found:

- PR #4162 fixes the signer-readiness race by delaying registration until a signer-backed Nostr session is ready.
- A prod-emulator test showed generic relay-pool publishing could report success while the kind `3079` event landed on a user relay (`nos.lol`) instead of `relay.divine.video`.
- A targeted diagnostic run against `wss://relay.divine.video` showed the relay rejected `3079` with `blocked: kind 3079 not in allowed list`.
- Relay-side allowlist/discovery work is tracked in `divine-funnelcake#380`.
- First-party mobile does not publish kind `1` events today.
- Mobile consumes kind `1` for mention-style notifications.
- Mobile comments/replies are currently published as kind `1111`.
- This issue should not expand mobile push preferences to `1111` or `30023`.

## Scope
- Publish push control events directly to `EnvironmentConfig.relayUrl`.
- Await NIP-20 relay `OK` for push control events.
- Use a bounded `5s` relay OK timeout.
- Log relay outcomes:
  - `acceptedBy`
  - `rejectedBy`
  - `noResponseFrom`
- Apply targeted relay + OK behavior to:
  - `3079` push registration.
  - `3080` push deregistration.
  - `3083` push preferences.
- Preserve the readiness-driven registration lifecycle from PR #4162.
- Preserve token-refresh re-registration behavior, but make refreshed-token registration use the same targeted relay + OK path.
- Keep `p` tags targeting `EnvironmentConfig.pushServicePubkey`.
- Keep existing `NotificationPreferences` trigger kind list unchanged for now: `[1, 3, 7, 16]`.
- Add temporary relay-layer diagnostics for push control kinds `3079/3080/3083` during rollout.
- Keep permanent push-control logs in `PushNotificationService`.
- Do not add a durable retry/background scheduler in this issue.

## Current Mobile Kind Contract Notes
- `1`: consumed for mention-style notifications; not first-party published by mobile; not current comments/replies.
- `3`: follows/contact-list.
- `7`: likes/reactions.
- `16`: generic reposts.
- `1111`: current mobile comments/replies, but not yet advertised in `3083` preferences.
- `30023`: mobile can consume long-form mentions locally, but not yet advertised in `3083` preferences.

## Out of Scope
- Adding `1111` to mobile push preferences.
- Adding `30023` to mobile push preferences.
- Changing canonical notification trigger-kind policy.
- Adding relay-side allowlist changes.
- Adding durable retry queues or background retry scheduling.
- Hardening relay policy to require `p` tags for push control kinds.

## Acceptance Criteria
- [ ] `3079` registration targets `EnvironmentConfig.relayUrl` and awaits/logs relay OK/reject/no-response.
- [ ] Token-refresh registration uses the same targeted relay + OK handling.
- [ ] `3080` deregistration targets `EnvironmentConfig.relayUrl` and awaits/logs relay OK/reject/no-response.
- [ ] `3083` preferences targets `EnvironmentConfig.relayUrl` and awaits/logs relay OK/reject/no-response.
- [ ] Push control publishes use a bounded `5s` relay OK timeout.
- [ ] Current `NotificationPreferences.toKindsList()` remains `[1, 3, 7, 16]` by default.
- [ ] No new durable retry/background scheduler is introduced.
- [ ] Temporary relay-layer diagnostics are limited to push control kinds `3079/3080/3083`.

## Test Plan
- Unit test `3079` registration targets `EnvironmentConfig.relayUrl`.
- Unit test `3079` registration awaits relay OK and handles accepted outcome.
- Unit test `3079` registration handles relay rejection without throwing and logs rejection reason.
- Unit test `3079` registration handles relay no-response timeout without throwing and logs `noResponseFrom`.
- Unit test token-refresh registration uses targeted relay + OK handling.
- Unit test `3080` deregistration targets `EnvironmentConfig.relayUrl`.
- Unit test `3080` deregistration awaits relay OK and handles accepted/rejected/no-response outcomes.
- Unit test `3083` preferences target `EnvironmentConfig.relayUrl`.
- Unit test `3083` preferences await relay OK and handle accepted/rejected/no-response outcomes.
- Unit test current `NotificationPreferences.toKindsList()` remains `[1, 3, 7, 16]` by default.
- Manual prod-emulator verification after `divine-funnelcake#380` is deployed:
  - Sign in with an allowlisted key.
  - Confirm `3079` receives `OK true` from `wss://relay.divine.video`.
  - Confirm `divine-push-service` observes registration.
  - Confirm backend stores token metadata.
  - Confirm registration does not rely on user relays such as `nos.lol`.
  - Confirm logs show target relay, relay OK result, and push-service processing evidence.

## Dependencies
- Depends on `divine-funnelcake#380` or equivalent deployment allowing push control kinds on `relay.divine.video`:
  - `3079`
  - `3080`
  - `3083`
- `divine-funnelcake#380` should also advertise newly allowed control kinds in discovery metadata.
- `divine-push-service` should continue to validate/decrypt/process:
  - `3079` registration.
  - `3080` deregistration.
  - `3083` preferences.
- Relay `OK true` only proves relay acceptance/storage/broadcast. It does not prove `divine-push-service` decrypted or persisted the registration, so manual verification must check both relay OK and push-service registration storage/logs.
- Future `divine-push-service` contract cleanup should decide:
  - Whether kind `1` should remain a push trigger for text-note mentions.
  - Whether comments/replies should use kind `1111` as a push trigger.
  - Whether long-form mentions should use `30023`.
  - Whether/when mobile should advertise `1111` and/or `30023` in `3083` preferences.

### PR DESCRIPTION
**Pull request by dcadenas**
## Summary

Push registration could look successful even when the Divine relay never accepted the event.
This change sends push registration, deregistration, and preference events directly to the configured Divine relay and waits for a relay OK response.
It also keeps the publish bounded so a late relay connection cannot write or queue a stale push-control event after the timeout.

## What Changed

Push control events now use the same targeted OK-confirmed publish path.
The relay layer now treats the caller timeout as one deadline for both sending and waiting for OK responses.
Same-URL fallback now distinguishes a failed send from a sent frame, and deadline-bound sends cannot queue stale EVENT messages.

- Route kinds 3079, 3080, and 3083 through `publishEventAwaitOk` with `targetRelays: [EnvironmentConfig.relayUrl]`.
- Let explicit target-relay OK publishes run even when the general relay pool has no connected relays.
- Skip the generic disconnected-relay retry when explicit target relays are provided.
- Bound `RelayPool.sendEventAwaitOk` with one deadline across `_sendCollect` plus OK waiting.
- Thread that deadline through `Relay.send`, `RelayBase.send`, and `WebSocketConnectionManager.sendJson`.
- Disable pending-message queueing for deadline-bound await-OK sends.
- Do not queue pending-auth messages for deadline-bound await-OK sends.
- Remove expired temp relays after deadline timeout so their connection waiters cannot flush stale control events.
- Track expected relays as soon as each send succeeds, so timeouts report the relays that actually received the frame.
- Avoid retrying a stalled same-URL normal target as a temp relay after the shared deadline is consumed.
- Allow same-URL temp fallback when the normal target path returns `false` immediately without sending.
- Log accepted, rejected, and no-response relay outcomes without logging encrypted push-control content.

**Related Issue:** Closes #4411

## Testing

I tested the push-control service paths, the client OK-publish wrapper, and the relay-pool timeout and fallback behavior.
I also verified the touched packages and the app analyzer after the SDK API change.

- `dart format packages/nostr_client/lib/src/nostr_client.dart packages/nostr_client/test/src/nostr_client_test.dart packages/nostr_sdk/lib/relay/relay_pool.dart packages/nostr_sdk/test/unit/relay_pool_concurrency_test.dart`
- `dart format --output=none --set-exit-if-changed lib/relay/relay.dart lib/relay/relay_base.dart lib/relay/relay_isolate.dart lib/relay/web_socket_connection_manager.dart lib/relay/relay_pool.dart test/unit/relay_pool_concurrency_test.dart test/unit/relay_pool_auth_test.dart test/unit/relay_count_test.dart test/integration/relay_auth_test.dart test/integration/nip50_search_test.dart`
- `dart format --output=none --set-exit-if-changed test/src/follow_repository_test.dart`
- `flutter test --no-pub test/src/nostr_client_test.dart --plain-name "publishEventAwaitOk"`
- `flutter test --no-pub test/unit/relay_pool_concurrency_test.dart --plain-name "sendEventAwaitOk"`
- `flutter test --no-pub test/unit/relay_pool_concurrency_test.dart`
- `dart test test/unit/publish_outcome_test.dart`
- `flutter test --no-pub test/services/push_notification_service_test.dart`
- `flutter test --no-pub test/providers/push_notification_sync_test.dart`
- `dart analyze` in `mobile/packages/nostr_sdk`
- `flutter analyze --no-pub` in `mobile/packages/nostr_client`
- `flutter analyze --no-pub` in `mobile/packages/follow_repository`
- `flutter analyze --no-pub` in `mobile`
- `flutter test --no-pub` in `mobile`
- `git diff --check`

## Risks

The main behavior change is that explicit target publishes no longer wait for unrelated pool reconnects, and deadline-bound OK publishes no longer queue messages for later delivery.
That is intentional for push control events because a late queued control event can make the app report a failed publish while still mutating relay-side push state later.
The SDK `Relay.send` interface now accepts an optional deadline, so custom test relays and package-local fake relays were updated to match.

- A target URL already present in the normal relay pool is not retried as a temp relay after the normal path sends or queues the frame.
- A normal target path that returns `false` immediately can still fall back to a same-URL temp relay within the caller timeout.
- If the normal target path stalls until the shared deadline is exhausted, the publish fails fast instead of trying the same URL a second way.
- If a relay connects after the deadline, the deadline-aware send path returns false before writing the stale EVENT.
- This does not prove the push service processed the event after relay acceptance; it only confirms relay acceptance.

## Type of Change

- [ ] ✨ New feature (non-breaking change which adds functionality)
- [x] 🛠️ Bug fix (non-breaking change which fixes an issue)
- [ ] ❌ Breaking change (fix or feature that would cause existing functionality to change)
- [ ] 🧹 Code refactor
- [ ] ✅ Build configuration change
- [ ] 📝 Documentation
- [ ] 🗑️ Chore

### CODE REVIEW

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart [hunk 1]
```diff
@@ -105,6 +105,31 @@ class RelayPool {
   List<MapEntry<String, Relay>> _cacheRelayEntriesSnapshot() =>
       _cacheRelays.entries.toList(growable: false);
 
+  bool _isPushControlKind(int? kind) =>
+      kind == 3079 || kind == 3080 || kind == 3083;
```
Comment (realmeylisdev) [inline mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: **SDK layering violation.** `architecture.md` is explicit:

> The data layer should be free of any specific domain or business logic. Packages within the data layer could be plugged into unrelated projects that need to retrieve data from the same sources.

The kind triplet `[3079, 3080, 3083]` is Divine's push-service contract — `nostr_sdk` should not know they exist. The same triplet is also duplicated on `PublishTracker.isPushControlEvent` in `publish_outcome.dart:83`, and `PushNotificationService` already names them as `pushRegistrationKind` / `pushDeregistrationKind` / `pushPreferencesKind`. Three sources of truth for one triplet.

A reusable design that achieves the same observability outcome without baking domain knowledge into the SDK:

```dart
Future<PublishOutcome> sendEventAwaitOk(
  List<dynamic> message, {
  required String eventId,
  int? eventKind,
  List<String>? tempRelays,
  List<String>? targetRelays,
  Duration timeout = const Duration(seconds: 15),
  String? diagnosticTag, // ← caller-supplied, e.g. "📲 push-control"
}) async { ... }
```

Then `PushNotificationService._publishPushControlEvent` passes `diagnosticTag: '📲 push-control'`, and `_sendCollect` emits the diagnostic block iff `diagnosticTag != null`. Same observability, no layering violation. Most of the scattered `if (isPushControlEvent) { _logPushControlDiagnostic(...) }` checks collapse to a single `if (diagnosticTag != null)` per fan-out branch.

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart [hunk 2]
```diff
@@ -105,6 +105,31 @@ class RelayPool {
...
+  String? _eventIdFromMessage(List<dynamic> message) {
+    if (message.length < 2 || message[0] != 'EVENT' || message[1] is! Map) {
+      return null;
+    }
+    final id = (message[1] as Map)['id'];
+    return id is String ? id : null;
+  }
+
+  void _logPushControlDiagnostic(String message) {
+    log('📲 Push control relay diagnostic: $message');
```
Comment (realmeylisdev) [inline mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: **Missing TODO + tracking issue for temporary code.** `code_style.md` requires:

> Transitional or temporary code (feature flags, compatibility shims, workarounds for in-progress migrations) must include a `// TODO(#issue):` comment referencing a tracking issue for its removal.

Issue #4411's scope explicitly says: *"Add **temporary** relay-layer diagnostics for push control kinds 3079/3080/3083 during rollout."* Without a TODO + tracking issue, the `📲 Push control relay diagnostic:` scaffolding becomes permanent dead weight — exactly the failure mode `agent_workflow.md` §4 ("Do not accumulate technical debt") exists to prevent.

Suggest opening a tracking issue (e.g. *"Remove push-control diagnostic wiring after `divine-funnelcake#380` is verified in production"*) and adding `// TODO(#xxxx):` markers at the call sites once the SDK refactor in the previous comment is in place.

File: mobile/packages/nostr_sdk/lib/relay/publish_outcome.dart [hunk 3]
```diff
@@ -18,6 +19,9 @@ class PublishOutcome {
   /// The id of the event that was published.
   final String eventId;
 
+  /// The event kind that was published, when known.
+  final int? eventKind;
```
Comment (realmeylisdev) [inline mobile/packages/nostr_sdk/lib/relay/publish_outcome.dart:27]: **Public API contract worth a one-line dartdoc.** `eventKind` is now public on `PublishOutcome`, but it's only populated when `RelayPool.sendEventAwaitOk` callers pass an explicit `eventKind` or the message envelope can be parsed. Other publish paths through `NostrClient` (e.g. `publishEvent` returning `PublishResult` rather than `PublishOutcome`) don't surface kind, so a future consumer assuming this is always populated will write null-blind code that breaks when called from those paths.

Suggest clarifying: *"Set by `RelayPool.sendEventAwaitOk` callers; null when older publish paths are used or when the event kind cannot be inferred from the message envelope."*

File: mobile/packages/nostr_sdk/lib/relay/publish_outcome.dart [hunk 4]
```diff
@@ -71,6 +76,12 @@ class PublishTracker {
   /// The event id we are waiting on.
   final String eventId;
 
+  /// The event kind we are waiting on, when known.
+  final int? eventKind;
+
+  bool get isPushControlEvent =>
+      eventKind == 3079 || eventKind == 3080 || eventKind == 3083;
```
Comment (realmeylisdev) [inline mobile/packages/nostr_sdk/lib/relay/publish_outcome.dart]: **Same SDK layering violation as `relay_pool.dart:109`.** Divine-specific kind numbers (`3079/3080/3083`) in a generic `nostr_sdk` package — and a duplicate of the literals on `RelayPool._isPushControlKind`. Goes away once `sendEventAwaitOk` accepts a `String? diagnosticTag` parameter from the caller (see comment on `relay_pool.dart`).

File: mobile/packages/nostr_sdk/lib/relay/relay_pool.dart [hunk 5]
```diff
@@ -943,12 +1028,34 @@ class RelayPool {
...
+              relay.pendingAuthedMessages.add(message);
+              sentTo.add(relay.url);
+              attemptedRelayUrls.add(relay.url);
+              onSent?.call(relay.url);
+              if (isPushControlEvent) {
+                _logPushControlDiagnostic(
+                  'queued for auth kind=$eventKind eventId=$eventId relay=${relay.url}',
+                );
+              }
+            } else if (isPushControlEvent) {
```
Comment (realmeylisdev) [inline mobile/packages/nostr_sdk/lib/relay/relay_pool.dart]: **Behavioural edge case worth a one-line doc comment.** When `deadline != null` and a relay requires AUTH (and isn't `vine.hol.is`), the message is silently dropped — neither queued for after-AUTH delivery nor sent. For push-control this is the correct contract (5s timeout, no point queueing a frame that will publish after the caller has already returned a failure outcome), but a future deadline-bound caller hitting this path will wonder why their event vanished without a trace.

Suggest adding a one-liner above the `if (deadline == null)` branch (line ~1042): *"Deadline-bound sends intentionally do not queue for AUTH: a queued frame would publish after the caller has already returned a failure outcome, mutating relay state behind the caller's back."*

**Review by realmeylisdev — COMMENTED** [review]
Net positive — the deadline plumbing, `onSent` tracker-race narrowing, same-URL fallback semantics, and `removeExpiredTempRelay` are all well-thought-out, and tests cover the key edge cases. The fix achieves issue #4411's acceptance criteria.

Two architectural concerns I'd want addressed before merge, both inline:

1. **SDK layering** — Divine-specific kind numbers (`3079/3080/3083`) leak into the domain-agnostic `nostr_sdk` package via `_isPushControlKind`, `isPushControlEvent`, and `_logPushControlDiagnostic`. See the inline on `relay_pool.dart:109` for a `diagnosticTag` redesign that keeps the SDK generic and collapses most of the scattered `if (isPushControlEvent) { ... }` checks.
2. **Temporary code without TODO + tracking issue** — issue #4411 explicitly scopes these as *temporary* diagnostics for the rollout, but `code_style.md` requires `// TODO(#xxxx):` markers tied to a removal-tracking issue. Without one they become permanent dead weight (see `agent_workflow.md` §4).

Plus two smaller items inline (auth-queue silent-skip doc comment, `PublishOutcome.eventKind` contract dartdoc).

## Traced error propagation chain

`PushNotificationService._publishPushControlEvent`
→ `NostrClient.publishEventAwaitOk(targetRelays: [relayUrl])`
→ `Nostr.sendEventAwaitOk` (now passes `eventKind`)
→ `RelayPool.sendEventAwaitOk` (registers tracker, calls `_sendCollect` with `deadline = now + timeout`)
→ `_sendCollect` (sequential fan-out, threads `deadline` through every `relay.send`)
→ `RelayBase.send` / `RelayIsolate.send` → `WebSocketConnectionManager.sendJson` → `send` → `_doSend`
→ each layer rechecks `_deadlineExpired(deadline)` before writing or queueing

The deadline plumbing is solid; my concerns are about the diagnostic scaffolding shape, not the behavioural fix.

## What's well done

- Deadline rechecks at every layer (reconnect wait, connecting wait, sink write, queue path) — `_remainingOr` and `_connectionWaitTimeout` correctly clamp to residual deadline.
- Switching from `tracker.expectedRelays.addAll(sentTo)` (after `_sendCollect` returns) to a synchronous `onSent` callback closes a real race window where a fast-responding relay could complete the tracker against an empty `expectedRelays` set.
- `attemptedRelayUrls.add(relay.url)` only on `result == true`, so the temp-relay loop *can* still try the same URL when the normal-pool send returned `false` immediately, but *won't* retry it after a stalled send consumed the shared deadline. Both behaviours pinned by tests.
- `removeExpiredTempRelay` prevents a phoenix temp relay from connecting after the deadline and flushing a stale push-control frame later.
- Cache rollback semantics preserved in the explicit-targets branch.

## Alternative architectures I evaluated

| Approach | Verdict |
|---|---|
| **Repository-layer fallback** (per `architecture.md`'s "fallback belongs in the repository") | Cleanest in theory, but introducing a `PushControlRepository` for one one-line wrapper is overkill. |
| **Service-layer wrapper** (this PR) | Right call. `_publishPushControlEvent` lives on `PushNotificationService`, which already composes all three flows (register, deregister, prefs). |
| **Add `publishPushControlEvent` to `NostrClient`** | Wrong — pushes Divine domain knowledge into `nostr_client`, the violation we're flagging in #1. |

The service-layer wrapper is right; the cleanup is on the SDK side, not the service side.

**Comment by realmeylisdev** [comment]
CI is crying a little bit, and rest of it LGTM!

**Review by NotThatKindOfDrLiz — APPROVED** [review]
Reviewed the updated head 6a9defa37d8dd31747fc0dc8ed96121f7ca5abd5. The target-send timeout race is fixed, the same-URL temp fallback behavior is now deterministic, the stale review threads are resolved, and the branch is green.

**Review by NotThatKindOfDrLiz — APPROVED** [review]
All checks pass. Blocking race fixed, stale threads resolved, history cleaned, CI green on 6a9defa37. Ready to merge.
